### DISCUSSION TITLE
Phazerville Suite v1.6.x

### DISCUSSION THREAD

**Discussion by djphazer** [djphazer/O_C-Phazerville#14] [Announcements]
General discussion space for the current v1.6.x series of my firmware. Brainstorming is very welcomed!

Refer to the [release notes](https://github.com/djphazer/O_C-BenisphereSuite/releases) for details and updates.
And be sure to read the Wiki!

**Comment by anxst** [comment]
Not worth a pull request, but you may want to change the documentation in the 1.6 release section (and forward) from:

> All of the original O_C apps (sans References) are in here now, available in 2 alternative builds.
> All 3 builds include Hemisphere, Low-rents, and Pong, plus:
> 
> - Main builds = Calibr8or, Quadraturia, Piqued, Enigma, Captain MIDI, Neural Net, Darkest Timeline
> - The "+stock" builds both include: CopierMaschine, Quantermain, Acid Curds, Automatonnetz, and:
>   - +stock1 = Piqued, Sequins
>   - +stock2 = Quadraturia, Meta-Q, Harrington 1200, Dialectic Ping Pong, Viznutcracker

to:

```
All of the original O_C apps (sans References) are in here now, in  all 3 alternative builds.
All builds include Hemisphere, Low-rents, and Pong. The build differences are:

- Main builds include Calibr8or, Quadraturia, Piqued, Enigma, Captain MIDI, Neural Net, Darkest Timeline

Stock builds instead include:
- +stock1 = CopierMaschine, Quantermain, Acid Curds, Automatonnetz, Piqued, Sequins
- +stock2 = CopierMaschine, Quantermain, Acid Curds, Automatonnetz, Quadraturia, Meta-Q, Harrington 1200, Dialectic Ping Pong, Viznutcracker

```
It's subtle, but a number of us were confused, and it lead to some discussion on Discord, and loading up different firmwares to figure out what was where. Thanks to Tom Churchill for working out what seemed (to a few of us) the best wording. Just a suggestion of course, and thank you for all your hard work on this!

**Reply by djphazer** [reply]
Ahh thank you! Communication is difficult sometimes 🙃

**Comment by athousanddetails** [comment]
@djphazer love this release man! Thanks for all your hard work. 

I was trying to see the code for SequencerX, would be dope to add Double Encoder Press = Randomize the Sequencer.

My C++ coding skills are super limited to attempt such thing! Cheers man

**Reply by djphazer** [reply]
This is a neat idea! I may attempt it... but I might add a button rather than using encoder double-press. It would suck to accidentally randomize your sequence while editing!

**Comment by Burdienco** [comment]
Good day! Wanted to ask about costum boundle of applets - if I want to have for example build stok 2, where quadraturia is, but instead of any random applet (Harrington 1200) insert Piqued. So my goal is not to brake something in existing build, but to have quadraturia and Piqed in one pack. I am without programming expirience, but may be this is not the most difficult task to dial with. And for now I am not shure where to start, any advice would be greatly appreciated.

**Reply by djphazer** [reply]
FWIW, Quadraturia and Piqued are both already in the main builds!

I may try to tackle a custom build system soon, but in the meantime, if you can get Platform IO installed, custom App selections can be made by editing the `platformio.ini` file. I have my own custom config in there, labeled "pewpewpew" which you can build with this command:

```
pio run -e pewpewpew
```

**Comment by mebitek** [comment]
could we select the scale for enigma jr?

**Reply by djphazer** [reply]
tbh, I never quite understood the way that works... I think you set up the scale settings in the full-size Enigma app, and Enigma Jr. is more like a melody player. One of these days, maybe I'll dive deeper into it.

**Reply by mebitek** [reply]
I was looking at the code and I think is possibile to add a menu entry to select the scale and then quantise the final output of enigmajr. howver you are saying that is possible to select the scale for each register in full enigma app?

**Reply by scuzzphut** [reply]
first thing Synthdad says in this video

https://www.youtube.com/watch?app=desktop&v=zt8Xm_3qOFc

**Comment by cuneytozseker** [comment]
Thanks for the build! Quick question: On the DualTM page it says Scale / Root Note but I cannot find the option to change the root note, only the scale?

**Reply by djphazer** [reply]
Root Note was just added in v1.6.6 !

**Reply by cuneytozseker** [reply]
Wow! Great timing!

**Comment by scuzzphut** [comment]
Firstly - thank you so much for your work on this.  I have 3 O&Cs all running Phazerville now, including a Teensy 4.0 version.

I am seeing an issue in the DualTM app.  When I adjust the probability parameter using the panel encoder, it acts as expected and bits are flipped in the sequence according to it's value.   When I assign the probability to a CV input and adjust it using a CV signal, there is no change to the sequence at all.  I can leave it at max probability for minutes and the sequence does not change. 

EDIT  -  Ignore the above guff. I was forgetting to unlock the probability with the gate on T2.  Must try harder. 

Thanks again.

**Comment by jpbed1** [comment]
Incredibly dumb question here, but for the (benispheres mode) screensaver, I noticed that the display never goes fully blank.

Is it supposed to, or is there any way to make it so?

**Reply by djphazer** [reply]
Ah, not currently. I replaced the blank screen with that screensaver, but I can potentially implement a secondary blank screen state. Or maybe just a config flag to disable the screensaver.

**Comment by jpbed1** [comment]
Well I broke down and did it :) Got up and running with Platform IO via VSCode, and just went in and commented out the current screensaver in the 777 release code. I wouldn't say it blanks "properly" now, but it certainly blanks!

**Reply by djphazer** [reply]
Heyy, right on! Now you're a firmware developer!

I've actually been working on new screensaver modes, including one that's just blank. Try out the code on the `dev/1.6` branch 😉

More discussion on #34

**Comment by TricksterSam** [comment]
Do you welcome some changes/additions?

In a previous version of Benisphere, I changed the Runglbook to have control over bit depth. I also changed the second cv output to be an xor. This allows for the easy creation of a Benjolin without needing to use the second app for the xor. I'm thinking about completing this by allowing the user to select what the second cv outputs.

I also developed a [Relabi app](https://github.com/TricksterSam/O_C-BenisphereSuite_HEM_Relabi/blob/main/software/o_c_REV/HEM_Relabi.ino) based on the work I've done with John Berndt on generative methods for [relabi](https://johnberndt.org/relabi/). It required an additional SetPhase function in WaveformManager.h.

`    void SetPhase(uint8_t index) {
        segment_index = index;
        signal = scale_level(segments[segment_count - 1].level);
        rise = calculate_rise(segment_index);
        sustained = 0;
        eoc = !cycle;
    }`

The Relabi app generates two LFOs signals that do not precisely repeat. It's great with comparators to create wonky trigger patterns or to create slow moving, unpredictable CV motion.

My eventual goal is to allow two Relabi apps to extend the same generative function across both hemispheres sharing the underlying LFO functions. I'm sort of a hack programmers who is learning a bunch right now, but it all works. In fact, with Phazerville, I'm no longer seeing any weird screen glitches with my fork!

Today, I added my Relabi app a downloaded version of Phazerville, but I haven't uploaded it yet. You can still download and use the Relabi app if you add it to hemisphere_config.h and add the SetPhase function to WaveformManager.h.

If anyone has suggestions about improvements, I'm open to all of your feedback. Maybe, I could do the phase reset another way from within the app. Or maybe there's a different way to do it.

Anyway, please test out the relabi app.

**Reply by djphazer** [reply]
Awesome! I'll definitely be trying this stuff out soon!

If you didn't see, I took a moment to rebase your work on the fork with existing git histories - https://github.com/TricksterSam/O_C-BenisphereSuite_RunglrRevised/pulls
So, this is the merged branch - https://github.com/djphazer/O_C-BenisphereSuite/tree/relabi-phz

I understand you're still learning, so I'm trying to leave breadcrumbs ;)
I may have some fixes & improvements for you when I actually test and examine the code.

**Reply by TricksterSam** [reply]
Oh, that's really helpful! Thanks for the merged branch. I'll try to set it up with Visual Studio before I go any further.

**Comment by dvvii** [comment]
Just got around trying your neat builder and I'm excited to start delving into the new goodies in here!

I was wondering why the Calibr8 applet disappeared from Hemispheres? I've been using it to tune two oscillators in precise intervals for FM synthesis—especially nice to do with Ebb&Flow since it gives precise tuning in hz. I was thinking it could be useful and perhaps not too hard to make Calibrat8 display the tuning offset in terms of ratios as is commonly approached for FM. I thought I'd mention it here since it's been removed as a motivation to bring it back? It could be just in a more simplified way for this purpose alone. I'm happy to chip in for more Ko-Fi!

In any case, thanks for all your work!

**Reply by djphazer** [reply]
OK, I'll resurrect the Calibr8 applet! I had removed it to save space, assuming it wasn't that useful...
For Ebb&LFO, you would get more precise tuning by simply adjusting the frequency in the applet ;P

Thanks for the Ko-Fi!

**Comment by basswolfe** [comment]
Is the annular fusion applet removed
From the current hemispheres? I liked the interface, but I don't see it in my current version. I just got an ocp from plum.
I know there's the other euclidean applet that does pretty much the same thing. Did that replace it? 

This thing is amazing btw! Thank you for everything you've done. I'm definitely going to try the custom builds thing your set up

**Reply by djphazer** [reply]
Yep, I retired AnnularFusion in favor of the redesigned EuclidX. The former looked cool, but I think the latter is more efficient and more flexible.

**Reply by basswolfe** [reply]
that makes sense, thanks!

**Comment by TricksterSam** [comment]
I think either my calibration is off or I don't know which environment variables to use in my platformio.ini.

I have an After Later Audio Ornament & Crime, which I think uses the Teensy 3.1.

Using the Voltage app, what I'm seeing from O_c at 0V is a value of about -0.73. Setting O_c at about 0.75V is close to 0V.. 1.91V is close to 1V. 3V is close to 2V. -0.25V is close to -1V. At -1.33V it is about -2V.  -2.41 is close to -3V.

I'm using pewpewpew. Is this an issue with compiling code for Teensy 4.17.0 but running it on a Teensy 3.1? These are the flags that pewpewpew calls:

`[env]
platform = teensy@4.17.0
framework = arduino
board = teensy31
board_build.f_cpu = 120000000
lib_deps = EEPROM
build_flags = 
	-DTEENSY_OPT_SMALLEST_CODE
	-DUSB_MIDI
	-Iextern
build_src_filter = 
	+<*>
	-<resources/tideslite.cpp>
extra_scripts = pre:resources/progname.py
upload_protocol = teensy-cli`

It really looks like a weird scaling issue and not calibration. Am I missing something obvious?

**Reply by djphazer** [reply]
Strange... was the unit properly calibrated to begin with? The `pewpewpew` configuration should work for After Later, as long as it's not using VOR. Try stock Hemisphere or another known-good firmware, and if the problem is still there, you might need to calibrate the DAC.

BTW, v4.17.0 is just the version of the compiler toolchain, not the hardware. You can see "board" is properly set to "teensy31"

**Reply by djphazer** [reply]
Looking through your latest code now, in platformio.ini it looks like you were actually using the `wepwepvor` configuration, so that would explain weird output scaling on your non-VOR hardware...

**Reply by TricksterSam** [reply]
I switched to os_stock and I'm seeing much better results. 0V is very close to 0V. I'm making measurements through an ES-8 in Bitwig, so there could be other factors, but it mirrors what I'm seeing on a ZeroScope, too. -3V is about -2.91V in Bitwig. 0V is 0.04V. 5.25V is 4.97V. I think this will work.

I just remembered something that could be causing this to be miscalibrated. One of the protection resistors blew on my Ornament & Crime, and a friend of mine replaced it with the nearest match we could find. Am I right in thinking that could cause a small difference in output levels?

**Reply by TricksterSam** [reply]
Thank you for providing help. I think I've got it working well enough with pewpewpew now. There's a lot I have been figuring out in the O_c library that I can use.

HEMISPHERE_CENTER_CV and HEMISPHERE_3V_CV are really helpful!

Next thing might be implementing your new nifty encoder selector method.
